### PR TITLE
Document Client-Relay protocol

### PR DESCRIPTION
**Pull request by drortirosh**

### CODE REVIEW

File: docs/relay-deployment.md [hunk 1]
```diff
@@ -0,0 +1,148 @@
...
+## Compile the server
+
+On Linux:
+```
+yarn 
+make server
+cp build/server/bin/RelayHttpServer .
+```
+
+On Mac (using docker to build the linux image):
```
Comment (shahafn) [inline docs/relay-deployment.md]: Change "on mac" to "with docker", and add prerequisites on linux (Go, truffle etc.)

File: docs/relay-deployment.md [hunk 2]
```diff
@@ -0,0 +1,169 @@
+# Manual setup for a GSN relayer
+
+The following steps were tested on amazon AWS Ubuntu instance.
+
+Make sure to modify your network specific parameters: 
+
+Your hostname (instead of `example.com`), and infura `NETWORK` and `INFURATOKEN`.
+
+
+## Compile the server
```
Comment (spalladino) [inline docs/relay-deployment.md]: Agree on guiding users to compile the server instead of downloading it

File: docs/protocol.md [hunk 3]
```diff
@@ -0,0 +1,230 @@
+# The GSN Protocol
+
+This document comes to describe the protocol between the client application and the relay,
+ for the purpose of sending a transaction to the blockchain.
+
+Note that most of the client data is checked not by the relay but on-chain by the `RelayHub` contract.<br/> 
+The formal specification of the protocol is defined by [EIP1813](https://github.com/ethereum/EIPs/blob/master/EIPS/eip-1613.md)
```
Comment (nventuro) [inline docs/protocol.md]:
```suggestion
The formal specification of the protocol is defined by [EIP1613](https://github.com/ethereum/EIPs/blob/master/EIPS/eip-1613.md)
```

File: docs/protocol.md [hunk 4]
```diff
@@ -0,0 +1,230 @@
...
+## Overview
+
+At a high-level the client perform the following steps:
+
+1. [Create list of Potential Relays](#create-list)
+2. [Select a relay](#select-relay)
+3. [Create a request](#create-request)
+4. [Send Request to the Relay.](#send-request)
+5. [Validate Response.](#validate-response)
+6. [ Handle Relay Error Responses.](#relay-error)
```
Comment (nventuro) [inline docs/protocol.md]:
```suggestion
6. [Handle Relay Error Responses.](#relay-error)
```

File: docs/protocol.md [hunk 5]
```diff
@@ -0,0 +1,230 @@
...
+3. [Handling stuck transactions](#relay-underpriced-tx)
+
+## Client Detailed Request Flow
+<a name="create-list"></a> 
+### 1. Create a List of Potential Relays
+
+* First, the client MUST get the RelayHub used by the target contract, by calling `target.getHubAddr()`<br/>
+    **NOTE**: it is important that the client **doesn't assume** the target uses the global RelayHub address: If/when the RelayHub is updated, its the target contract's responsibility to update its hub link.
+* Next, the client SHOULD filter for recent `RelayAdded` and `RelayRemoved` events sent by the RelayHub.
+  Since a relay is required to send such event every 6000 blocks (roughly 24hours), the client should look at most at the latest 6000 blocks.
```
Comment (nventuro) [inline docs/protocol.md]: It'd be good to elaborate on the rationale behind this.

File: docs/protocol.md [hunk 6]
```diff
@@ -0,0 +1,230 @@
...
+  Ignore relays with stake time or stake delay below a given threshold. 
+* Sort the relays in the preferred order. e.g.: prefer relays with lower transaction fee, and also depend on the [Relay Reputation](#relay-reputation)
+* Dynamic relay selection: before making the call, the client SHOULD "ping" the relay (see below)
+* For each relay, the client keeps the relay `address` and `url`  
+
+<a name="select-relay"></a>
+###  2. Select a Relay
+
+* For each potential relay, the client "pings" the relay by sending a `/getaddr` request.
+* Validate the relay is valid (contains `valid:true`)
```
Comment (nventuro) [inline docs/protocol.md]: Should this be `Ready: true`?

File: docs/protocol.md [hunk 7]
```diff
@@ -0,0 +1,230 @@
...
+* Validate the relay supports this protocol: `version:1.0.x`
```
Comment (nventuro) [inline docs/protocol.md]: Is this advertised by the relayer?

File: docs/protocol.md [hunk 8]
```diff
@@ -0,0 +1,230 @@
...
+* The client SHOULD ping few relays, but not too much: e.g. default client pings 3 relays, and use the first valid one. 
+    Only if none of the first 3 relays answers, it will select the next 3 relays from the list  
+
+<a name="create-request"></a>    
+### 3. Create a Request
+
+* The client should create and sign a relay request, which MUST contain the following fields:
+  -  `from`: the client's address
+  -  `to`: the target contract's address
+  - `encodedFunction` - the function to call on the **target** contract.
```
Comment (nventuro) [inline docs/protocol.md]: I'd mention that this isn't just the selector, but also the calldata.

File: docs/protocol.md (hunk 7 above)
Comment (drortirosh) [reply docs/protocol.md]: tricky: the relay reports the *relay protocol* version, which is currently `0.4.1`
the relay-client protocol is not directly related to the RelayHub protocol, but it is heavily influenced by it (as  the client signs a struct that is checked by the RelayHub)

**Pull request merged**